components/codec/byte.rs

```diff
@@ -0,0 +1,1188 @@
+    /// and `written_bytes` is the number of bytes written.
+    ///
+    /// Note that actual written data may be larger than `written_bytes`. Bytes more than
+    /// `written_bytes` are junk and should be ignored.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::UnexpectedEOF` if `src` is drained while expecting more data.
+    ///
+    /// Returns `Error::BadPadding` if padding in `src` is incorrect.
+    ///
+    /// When there is an error, `dest` may contain partially written data.
+    #[inline]
+    pub fn try_decode_first_in_place(src: &mut [u8]) -> Result<(usize, usize)> {
+        let ptr = src.as_mut_ptr();
+        let len = src.len();
+        let src = unsafe { ::std::slice::from_raw_parts(ptr, len) };
+        let dest = unsafe { ::std::slice::from_raw_parts_mut(ptr, len) };
+        Self::try_decode_first(src, dest)
+    }
+
+    /// Decodes bytes in descending memory-comparable format in place, i.e. decoded data will
+    /// overwrite the encoded data.
+    ///
+    /// If there are multiple encoded byte slices in `src`, only the first one will be decoded.
+    ///
+    /// Returns `(read_bytes, written_bytes)` where `read_bytes` is the number of bytes read
+    /// and `written_bytes` is the number of bytes written.
+    ///
+    /// Note that actual written data may be larger than `written_bytes`. Bytes more than
+    /// `written_bytes` are junk and should be ignored.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::UnexpectedEOF` if `src` is drained while expecting more data.
+    ///
+    /// Returns `Error::BadPadding` if padding in `src` is incorrect.
+    ///
+    /// When there is an error, `dest` may contain partially written data.
+    #[inline]
+    pub fn try_decode_first_in_place_desc(src: &mut [u8]) -> Result<(usize, usize)> {
+        let ptr = src.as_mut_ptr();
+        let len = src.len();
+        let src = unsafe { ::std::slice::from_raw_parts(ptr, len) };
+        let dest = unsafe { ::std::slice::from_raw_parts_mut(ptr, len) };
+        Self::try_decode_first_desc(src, dest)
+    }
+
+    /// The internal implementation for both `try_decode_first` and `try_decode_first_desc`.
+    ///
+    /// It uses generics to specialize different code path for ascending and descending decoding.
+    /// Please refer to `try_decode_first` to see the meaning of return values, panics and errors.
+    fn try_decode_first_internal<T: MemComparableCodecHelper>(
+        src: &[u8],
+        dest: &mut [u8],
+        _helper: T,
+    ) -> Result<(usize, usize)> {
+        assert!(dest.len() >= src.len());
+
+        unsafe {
+            let mut src_ptr = src.as_ptr();
+            let mut dest_ptr = dest.as_mut_ptr();
+            let src_ptr_end = src_ptr.offset(src.len() as isize);
+
+            loop {
+                let src_ptr_next = src_ptr.offset((MEMCMP_GROUP_SIZE + 1) as isize);
+                if src_ptr_next > src_ptr_end {
```

Comment:
Do we need to decode multiple encoded byte slices in `src`?  otherwise, this branch can also be eliminated out the loop.